### ISSUE TITLE
Non-primary Tap Stream source names are changed after deploy, StackOverflow Exception is thrown when inspecting on dashboard

### ISSUE THREAD

**Issue by momogod** [spring-cloud/spring-cloud-dataflow#2150]
I initially created some tap streams as following:

Stream "base":
`workflow-service --server.port=8085 | log`   
Stream "comparator"
`:base.workflow-service > comparator-service | rabbit --routing-key=comparator --username=guest --password=******`
Stream "mapper"
`:base.workflow-service > mapper-one | rabbit --routing-key=mapper --username=guest --password=******            `

After few deploy and undeploy, non-primary tab streams are changed as following:
Stream "comparator"
`:comparator > comparator-service | rabbit --spring.rabbitmq.password=guest --rabbit.routing-key=comparator --spring.rabbitmq.username=guest  `
Stream "mapper" 
`:mapper > mapper-one | rabbit --spring.rabbitmq.password=guest --rabbit.routing-key=mapper --spring.rabbitmq.username=guest`

When inspecting the 'base' stream on the dashboard, it originally shows all the forks and tap streams. After changed, StackOverFlow exception was thrown:

> 2018-04-18 15:22:07.744 ERROR 32747 --- [nio-9393-exec-3] o.s.c.d.s.c.RestControllerAdvice         : Caught exception while handling a request
> 
> org.springframework.web.util.NestedServletException: Handler dispatch failed; nested exception is java.lang.StackOverflowError
> 	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:982) ~[spring-webmvc-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
> 	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901) ~[spring-webmvc-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
> 	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970) [spring-webmvc-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
> 	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861) [spring-webmvc-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
> 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:635) [tomcat-embed-core-8.5.27.jar!/:8.5.27]
> 
> Caused by: java.lang.StackOverflowError: null
> 	at java.util.Collections$UnmodifiableCollection.size(Collections.java:1030) ~[na:1.8.0_162]
> 	at org.springframework.cloud.dataflow.core.dsl.Tokens.hasNext(Tokens.java:131) ~[spring-cloud-dataflow-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.core.dsl.Tokens.next(Tokens.java:199) ~[spring-cloud-dataflow-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.core.dsl.StreamParser.eatDestinationReference(StreamParser.java:302) ~[spring-cloud-dataflow-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.core.dsl.StreamParser.eatSourceDestination(StreamParser.java:257) ~[spring-cloud-dataflow-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.core.dsl.StreamParser.eatStream(StreamParser.java:137) ~[spring-cloud-dataflow-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.core.dsl.StreamParser.parse(StreamParser.java:74) ~[spring-cloud-dataflow-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.server.controller.StreamDefinitionController.findRelatedDefinitions(StreamDefinitionController.java:296) ~[spring-cloud-dataflow-server-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.server.controller.StreamDefinitionController.findRelatedDefinitions(StreamDefinitionController.java:302) ~[spring-cloud-dataflow-server-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
> 	at org.springframework.cloud.dataflow.server.controller.StreamDefinitionController.findRelatedDefinitions(StreamDefinitionController.java:302) ~[spring-cloud-dataflow-server-core-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]

**Comment by sabbyanandan** [comment]
Hi, @momogod. Thanks for the report. The following 2 definitions don't look defined correctly.

```
║comparator │:base.workflow-service > comparator-service | rabbit --routing-key=comparator --username=guest --password=******
║mapper │:base.workflow-service > mapper-one | rabbit --routing-key=mapper --username=guest --password=******
```

A TAP can be at the beginning position, but not in between at the processor level. Both the Shell and UI should fail with errors.

Shell:
```
dataflow:>stream create bar --definition "time | filter --server.port=8085 | log"
Created new stream 'bar'
dataflow:>stream create bar1 --definition "http | :foo.filter > transform | rabbit --routing-key=comparator --username=guest --password=******"
Command failed org.springframework.cloud.dataflow.rest.client.DataFlowClientException: 118E:(pos 7): expected app name but found ':'
http | :foo.filter > transform | rabbit --routing-key=comparator --username=guest --password=******
    *  ^
```

UI:
![image](https://user-images.githubusercontent.com/1562654/38959279-e0f29370-4314-11e8-81e5-cea14a2a50d8.png)

How did it deploy in the first place? Maybe it is just formatted incorrectly in your description. If you can clarify this, it'd be helpful.

**Comment by momogod** [comment]
Hi Sabby,

The names behind the double-pipe are the stream names. The both actual streams start with ":base.workflow-service".

Sorry for the format of the copy&paste.

Thanks for the quick reply!
Charles

**Comment by momogod** [comment]
I've updated the original post with clearer context.

**Comment by sabbyanandan** [comment]
Thanks for the update! Could you also describe the following:

> After few deploy and undeploy, non-primary tab streams are changed as following

Is it from Shell/UI? If reproducible, what to do to replicate it?

**Comment by momogod** [comment]
Hi @sabbyanandan ,

Deploy and undeploy are through **Shell** and using Skipper mode. 

That StackOverflow Exception was thrown while inspecting the stream definition on **UI**. While clicking the "i" icon on one of the "non-primary tap streams".  

I am more concerned about "source name is changed". I think the infinite loop is caused by that. (maybe also validate the definition before actually find the related definition. such as not allow reference name the same as current stream name?)

I will try to provide some reproduce steps later. But on my local machine, it has happened 3 times. And no really special commands were sent other than deploy and undeploy. 

Cheers,
Charles

**Comment by momogod** [comment]
@sabbyanandan @aclement 

Here is the reproduce step:

1. create primary stream
2. create tap stream by using the same source from the primary stream.
3. deploy primary stream via Shell, with propertiesFile
4. deploy tap stream via Shell, with propertiesFile
5. run "stream list" and see the source of the tap stream is now changed. 

Hope it helps. 

Cheers,
Charles

**Comment by sabbyanandan** [comment]
Great info! Thank you, @momogod. We will look into it.

**Comment by aclement** [comment]
I've got a fix for the stackoverflow - it was related to the stream 'changing itself' so isn't really the root problem, but good to address anyway.

@momogod can you tell me what properties you tweak in your properties files for the deploy steps? @sabbyanandan  and i tried a basic variant of those steps above and it just worked, but I have my suspicions where the issue lies. We were just deploying very simple `time | log` type streams though, without a processor. It must be some combination of stream complexity (need a processor in there) and the properties files to trigger. I'd assume (hopefully?) that you didn't need to deploy the primary stream with properties in order to trigger this, since that stream doesn't get damaged, it is more about the tap stream being deployed with properties.

**Comment by momogod** [comment]
@aclement @sabbyanandan 

the properties are followings:
BASE:
app.workflow-service.spring.cloud.stream.bindings.output.destination=comparator,mapper
app.workflow-service.server.port=8085

COMPARATOR:
app.comparator-service.spring.cloud.stream.bindings.input.destination=comparator
app.comparator-service.spring.cloud.stream.bindings.output.destination=comparator
app.rabbit.spring.cloud.stream.bindings.input.destination=comparator


MAPPER:
app.mapper-one.spring.cloud.stream.bindings.input.destination=mapper
app.mapper-one.spring.cloud.stream.bindings.output.destination=mapper
app.rabbit.spring.cloud.stream.bindings.input.destination=mapper

**Comment by momogod** [comment]
@aclement @sabbyanandan 

Today I also notice another problem related to this.

1. Create another standalone stream, also using workflow-service (the same source in BASE)
2. undeploy all streams
3. only deploy new standalone stream with properties files. 
4. run "stream list" and see the source in the standalone stream is changed. same symptom.

**Comment by aclement** [comment]
I think it is the `app.comparator-service.spring.cloud.stream.bindings.input.destination=comparator` property, I think that is overriding the input destination for the comparator stream, replacing `:base.workflow-service` with `comparator` when the stream definition is rewritten in `StreamDefinitionToDslConverter`. Similarly the setting in the mapper properties is damaging the source of that stream on rewrite.

**Comment by momogod** [comment]
so it has nothing to do with tap stream? because it also happens on standalone stream, with similar settings

**Comment by sabbyanandan** [comment]
@momogod: Sorry, but I'm confused with what you're trying to accomplish by overriding destinations. 

Let me clarify something.

If you're using a destination directly at the source or sink position in the DSL, _it is_ actually the destination. In other words, that is the actual queue/topic you'd want to consume or produce data. 

In your case, if you want to consume from the topic `comparator`, your stream definition would be something like:

> :comparator > comparator-service | rabbit --routing-key=comparator --username=guest --password=*****

**Comment by momogod** [comment]
@sabbyanandan 

I am new to Spring Cloud, so I might have done some very funny configuration. :P

What I am trying to build is an application does following:

1. has a workflow-service as SOURCE that has a REST-API to accept request.
2. workflow-service dynamically send the request to different stream to process data.
3. workflow-service also listening the rabbit sink to get the result. return it back as REST Response. 

In short, a REST gateway that trigger SCDF streams to do processing and reply. 


so my question is.. why would deploy properties change the stream definition itself? It shouldn't do that because stream could be deployed to different environment with different config.

**Comment by sabbyanandan** [comment]
Let me clarify the connection between SCDF and "destinations". Consider the following stream.

> dataflow:>stream create for_momogod --definition "http | filter | cassandra

SCDF _creates_ well-defined properties _(i.e: stream_name + app_name)_ for all the apps that it deploys. You'd notice the following properties unless overridden.

_cassandra:_
> "spring.cloud.stream.bindings.input.destination": "for_momogod.filter"

_filter:_
> "spring.cloud.stream.bindings.input.destination": "for_momogod.http"
> "spring.cloud.stream.bindings.output.destination": "for_momogod.filter"

_http:_
> "spring.cloud.stream.bindings.output.destination": "for_momogod.http"

In your case, you're attempting to override the destinations. Because the "destinations" are defined in the starting position of a stream _(i.e a TAP position)_, the runtime overrides automatically takes precedence. Since the `stream list` command queries for the current runtime status of the apps, you're seeing the runtime reflection of the destination properties. 

Again, a [named-destinations](https://docs.spring.io/spring-cloud-dataflow/docs/1.5.0.M1/reference/htmlsingle/#spring-cloud-dataflow-stream-dsl-named-destinations) in SCDF corresponds to a actual queue/topic in the broker. If you intend to consume data from a queue/topic, use the actual queue/topic-name directly in the DSL as opposed to overriding it via the properties file.

**Comment by momogod** [comment]
Hi, @sabbyanandan 

Thank you for the very useful information. 

Maybe I misunderstood the intention of deploy properties and defining properties in the DSL. I thought:

>Stream definition is like a java class. 
>Deploy with properties is like creating an instance of that class. 
>Deployment shouldn't change the class definition at any time. 

However, stream changing itself is still very confusing and sometimes misleading. 

I thought having setup the definition, I can change some properties during the deployment phase. And when I need to change them, I can deploy the same TAP stream with some new properties. (without destroy and re-create the definition). 

------------------
I don't know if `stream list` is showing runtime status. Because when I check the database and the definition is modified. Even I shut down the SCDF server, Skipper and restart them, definition is still wrong. And if I deploy the wrong stream with different properties (different destination), it would throw error. (that said I cannot reuse the tap stream anymore)

Maybe it's not a bug, but it certainly could be a future improvement.

Thanks,
Charles

**Comment by momogod** [comment]
@sabbyanandan @aclement 

I really appreciate your knowledge sharing. I am now can finish my product prototype. 

Now by understanding the `:channel` is just replacing the incoming channel from source. I think I have a better control and understand how to config and design my workflow-engine.

Please let me know if you decide to fix the "stream changes itself" problem. Meanwhile, I start reading the source code. Hopefully I can contribute something to SCDF. 

Thanks a lot again,
Charles

**Comment by sabbyanandan** [comment]
@momogod: Thanks for the comments. Also, it is because of you, we were able to address the recursion problem, which is now fixed and merged to `master`. Kudos to you for finding it! 👏 

To create and deploy streams dynamically, perhaps there's an opportunity to leverage either programmatically via [Java DSL](https://docs.spring.io/spring-cloud-dataflow/docs/1.4.0.RELEASE/reference/htmlsingle/#spring-cloud-dataflow-stream-java-dsl) or it can be hacked using the [REST-APIs](https://docs.spring.io/spring-cloud-dataflow/docs/1.4.0.RELEASE/reference/htmlsingle/#api-guide-resources-stream-definitions) or [DataFlowTemplate](https://docs.spring.io/spring-cloud-dataflow/docs/1.4.0.RELEASE/reference/htmlsingle/#_using_the_data_flow_template). That could fit the dynamic data processing needs.

Anyways, feel free to reach out to us in StackOverflow or Gitter with more questions. And, yes, we look forward to your contributions! 😃
